### PR TITLE
Replace /admin/v1/users_paginate endpoint with /admin/v2/users

### LINKED ISSUE #5494: /admin/users_paginate is pointless and should be merged into /admin/users
#1784 introduced a (barely-functional, cf #3666) admin API called `/admin/users_paginate`. Apart from other problems with this API (#2522, lack of documentation), the whole thing seems pointless. How is it different to `/admin/users` ?

### PR DESCRIPTION
**Pull request by awesome-manuel**
Fixes #5494 and #2522.
Needs to be tested for #3666.

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.rst#changelog)
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.rst#sign-off)

### CODE REVIEW

**Comment by awesome-manuel** [comment]
@richvdh The pagination API is not 100% compatible to the normal users API. Pagination returns a list of users and the total amount as a separate field whereas the users API just returns a list.
We probably don't want to change the users API since this one is actually used by current clients.

**Comment by richvdh** [comment]
might be a case for `/admin/v2/users` which solves the various problems with the current endpoints.

**Comment by awesome-manuel** [comment]
@richvdh any idea why buildkite does not complete?

**Comment by awesome-manuel** [comment]
Can someone explain what this means exactly?
```
$ ./tools/check/check_code_quality.sh
Check drawable quantity
ERROR, missing drawable alternative.
```

File: changelog.d/5925.removal [hunk 1]
```diff
@@ -0,0 +1 @@
+Remove admin/v1/users_paginate endpoint. Contributed by Awesome Technologies Innovationslabor GmbH.
```
Comment (richvdh) [inline changelog.d/5925.removal]: ... and add pagination to `/users`

File: docs/admin_api/user_admin_api.rst [hunk 2]
```diff
@@ -1,3 +1,35 @@
...
+=============
+
+This API returns all local user accounts.
+
+The api is::
+
+    GET /_synapse/admin/v1/users?access_token=<admin_access_token>&start=0&limit=10
+
+including an ``access_token`` of a server admin.
+The parameters ``start`` and ``limit`` are optional and are used for pagination.
```
Comment (richvdh) [inline docs/admin_api/user_admin_api.rst]: convention among matrix pagination APIs (which I think we should copy here) is that pagination apis return a `next_token` value, which you then use for the next iteration. Also, the parameter for the starting point is normally called `from` rather than `start`.

Comment (richvdh) [inline docs/admin_api/user_admin_api.rst]: it should probably be limited by default.

**Comment by richvdh** [comment]
> any idea why buildkite does not complete?

buildkite doesn't run on contributed PRs until someone pushes the button. It's something we're working on fixing, but haven't yet.

**Comment by awesome-manuel** [comment]
> Sorry for the very slow review on this! I've got some suggestions, mostly around the shape of the API.
> 
> The current API is a disaster, so I'm treating this basically as a fresh start, hence some of my comments. But also: because some of the changes I'm asking for will be incompatible, I suggest it goes onto /admin/v2/users, rather than changing the existing endpoint.

I totally agree, but keeping the old API implies keeping some ugly methods in DataStore.

File: docs/admin_api/user_admin_api.rst (hunk 2 above)
Comment (awesome-manuel) [reply docs/admin_api/user_admin_api.rst]: This API uses offset paging instead of token based paging. See also https://www.moesif.com/blog/technical/api-design/REST-API-Design-Filtering-Sorting-and-Pagination/# for the limitation, but for a reasonable amount of users this should be OK.

File: synapse/rest/admin/v1/__init__.py [hunk 3]
```diff
@@ -0,0 +1,358 @@
+# -*- coding: utf-8 -*-
```
Comment (richvdh) [inline synapse/rest/admin/v1/__init__.py]: please don't do this; let's sort the modules by functionality, not version.

Comment (richvdh) [reply synapse/rest/admin/v1/__init__.py]: (also lumping a massive change like this into this PR makes it impossible to see what is actually changing)

File: docs/admin_api/user_admin_api.rst (hunk 2 above)
Comment (richvdh) [reply docs/admin_api/user_admin_api.rst]: Still my point remains: let's not use a different pagination system to the other matrix APIs.

I don't mind if the value that you return from `next_token` is the offset, so that it can be used in the same way.

**Comment by richvdh** [comment]
btw @awesome-manuel: it might be easier for you to drop into `#synapse-dev:matrix.org` if you'd like to discuss these changes?

**Comment by awesome-manuel** [comment]
> > My suggestion is to at least drop the pagination API from v1, since it never worked as intended anyway.
> 
> I'm happy with that.
> 
> I'm afraid I can't really review this as it stands, due to the size of the patch. Sorry - please can you undo the move into a `v1` package?

Yeah, that's a limitation of GitHub. You can have a look at each individual commit here: https://github.com/Awesome-Technologies/synapse/commits/user_admin
Each commit is self-contained (e.g. does not break anything). Of course you can't comment on each individual commit. Internally we use Gerrit, it's a huge improvement for the reviewer, though requires a bit more learning from the reviewee.

**Comment by erikjohnston** [comment]
@awesome-manuel I'm afraid I just merged #6196 which ports the `synapse.rest.admin` package to use async/await, which is the cause for all the conflicts. The change is fairly mechanical, `@defer.inlineCallbacks -> async def` and `yield -> await`, but I'm happy to resolve the merge conflict and push here if it turns out to be more difficult to resolve than that

File: synapse/rest/admin/v1/__init__.py (hunk 3 above)
Comment (awesome-manuel) [reply synapse/rest/admin/v1/__init__.py]: So this will change also for the client endpoints in the future?

**Comment by awesome-manuel** [comment]
> @awesome-manuel I'm afraid I just merged #6196 which ports the `synapse.rest.admin` package to use async/await, which is the cause for all the conflicts. The change is fairly mechanical, `@defer.inlineCallbacks -> async def` and `yield -> await`, but I'm happy to resolve the merge conflict and push here if it turns out to be more difficult to resolve than that

I think I fixed all of these. Can someone please trigger Buildkite?

File: synapse/rest/admin/v1/__init__.py (hunk 3 above)
Comment (richvdh) [reply synapse/rest/admin/v1/__init__.py]: maybe one day. But let's not copy the mistakes of `/rest/client` into `/rest/admin`.

File: synapse/rest/admin/users.py [hunk 4]
```diff
@@ -12,17 +12,394 @@
...
+    assert_user_is_admin,
+    historical_admin_path_patterns,
 )
-from synapse.rest.admin import assert_requester_is_admin, assert_user_is_admin
 from synapse.types import UserID
 
+logger = logging.getLogger(__name__)
+
+
+class UsersRestServlet(RestServlet):
```
Comment (richvdh) [inline synapse/rest/admin/users.py]: again: please don't move these (at least, not in the same PR). it's much harder to review the change when there are hundreds of lines of unrelated changes.

File: synapse/rest/admin/__init__.py [hunk 5]
```diff
@@ -342,45 +174,15 @@ def __init__(self, hs):
         return 200, purge_status.asdict()
 
 
-class DeactivateAccountRestServlet(RestServlet):
```
Comment (richvdh) [inline synapse/rest/admin/__init__.py]: again: please don't move the unchanged servlets (at least, not in the same PR). It's much harder to review the actual changes when the diff includes hundreds of lines of moving things around.

Comment (awesome-manuel) [reply synapse/rest/admin/__init__.py]: OK, I will put this change into a separate PR.

Comment (awesome-manuel) [reply synapse/rest/admin/__init__.py]: Done

**Comment by awesome-manuel** [comment]
What do you think about the `@deprecation` decorator to mark "old" functions?
https://pypi.org/project/deprecation/

File: docs/admin_api/user_admin_api.rst [hunk 6]
```diff
@@ -1,3 +1,45 @@
...
+The parameters ``offset`` and ``limit`` are required only for pagination.
+Per default a ``limit`` of 100 is used. If the endpoint returns less entries
+than specified by ``limit`` then there are no more users left.
+The parameter ``name`` can be used to filter by user name.
+The parameter ``guests`` can be used to exclude guest users.
+The parameter ``deactivated`` can be used to include deactivated users.
+
+It returns a JSON body like the following:
+
+.. code:: json
```
Comment (richvdh) [inline docs/admin_api/user_admin_api.rst]: this isn't coming out in the formatted file (https://github.com/Awesome-Technologies/synapse/blob/user_admin/docs/admin_api/user_admin_api.rst): possibly it needs an extra newline or something like that?

File: synapse/rest/admin/users.py [hunk 7]
```diff
@@ -59,72 +60,43 @@ def __init__(self, hs):
         return 200, ret
 
 
-class GetUsersPaginatedRestServlet(RestServlet):
-    """Get request to get specific number of users from Synapse.
+class UsersRestServletv2(RestServlet):
```
Comment (richvdh) [inline synapse/rest/admin/users.py]:
```suggestion
class UsersRestServletV2(RestServlet):
```

File: synapse/storage/_base.py [hunk 8]
```diff
@@ -1592,6 +1598,9 @@ def _simple_select_list_paginate_txn(
         Args:
             txn : Transaction object
             table (str): the table name
+            filters (dict[str, T] | None):
```
Comment (richvdh) [inline synapse/storage/_base.py]: I kinda feel that once we're adding `LIKE` clauses to the query, this is quite specialised, and it would make more sense to give `get_users_paginate` its own `txn` function which can run the sql itself.

File: synapse/storage/_base.py [hunk 9]
```diff
@@ -1606,10 +1615,12 @@ def _simple_select_list_paginate_txn(
         if order_direction not in ["ASC", "DESC"]:
             raise ValueError("order_direction must be one of 'ASC' or 'DESC'.")
 
+        where_clause = "WHERE " if filters or keyvalues else ""
+        if filters:
+            where_clause += " AND ".join("%s LIKE ?" % (k,) for k in keyvalues)
```
Comment (richvdh) [inline synapse/storage/_base.py]: s/keyvalues/filters/?

**Comment by richvdh** [comment]
> What do you think about the @deprecation decorator to mark "old" functions?
https://pypi.org/project/deprecation/

as a general principle: sure, sounds sensible. But I'm not quite sure how it applies here?

File: synapse/storage/_base.py (hunk 8 above)
Comment (awesome-manuel) [reply synapse/storage/_base.py]: Actually I guess some more endpoints with pagination could profit from this kind of query.

**Comment by richvdh** [comment]
could you avoid force-pushing please? it makes it hard to see what has changed since the last review.

File: synapse/storage/data_stores/main/stats.py [hunk 10]
```diff
@@ -260,6 +260,7 @@ def _get_statistics_for_subject_txn(
         slice_list = self._simple_select_list_paginate_txn(
             txn,
             table + "_historical",
+            None,
```
Comment (richvdh) [inline synapse/storage/data_stores/main/stats.py]: I think once we get to this many params, we should be using keyword args.

File: synapse/storage/_base.py (hunk 8 above)
Comment (richvdh) [reply synapse/storage/_base.py]: also: can you give the parameters default values so that we don't need to list them all out in `_get_statistics_for_subject_txn` ?

File: synapse/rest/admin/users.py [hunk 11]
```diff
@@ -59,72 +60,43 @@ def __init__(self, hs):
...
 
-        order = "name"  # order by name in user table
-        params = parse_json_object_from_request(request)
-        assert_params_in_dict(params, ["limit", "start"])
-        limit = params["limit"]
-        start = params["start"]
-        logger.info("limit: %s, start: %s", limit, start)
+        start = parse_integer(request, "offset", default=0)
+        limit = parse_integer(request, "limit", default=100)
+        name = parse_string(request, "name", default=None)
```
[the remark this answers is not in the document]
Comment (richvdh) [reply synapse/rest/admin/users.py]: sorry, yes it is

File: docs/admin_api/user_admin_api.rst [hunk 12]
```diff
@@ -1,3 +1,46 @@
...
+
+The api is::
+
+    GET /_synapse/admin/v2/users?offset=0&limit=10&guests=false
+
+including an ``access_token`` of a server admin.
+The parameters ``offset`` and ``limit`` are required only for pagination.
+Per default a ``limit`` of 100 is used. If the endpoint returns less entries
+than specified by ``limit`` then there are no more users left.
+The parameter ``name`` can be used to filter by user name.
```
Comment (richvdh) [inline docs/admin_api/user_admin_api.rst]: it seems to be filtering by user_id rather than by (display) name. The `users` table has a `name` column, but that's a historical error; we refer to user_id elsewhere.

Comment (richvdh) [inline docs/admin_api/user_admin_api.rst]: can you document the format of the parameter? is it a regular expression or does it have to match exactly?

**Comment by awesome-manuel** [comment]
> could you avoid force-pushing please? it makes it hard to see what has changed since the last review.

Agreed. Would you prefer a fixup commit upon the existing ones, or changing the relevant commits but without rebase?

File: docs/admin_api/user_admin_api.rst [hunk 13]
```diff
@@ -1,3 +1,46 @@
...
+The parameter ``guests`` can be used to exclude guest users.
```
Comment (richvdh) [inline docs/admin_api/user_admin_api.rst]: can you document the acceptable values for this? What should I set it to to exclude guest users? Can I use it to return *only* guest users?

File: docs/admin_api/user_admin_api.rst [hunk 14]
```diff
@@ -1,3 +1,46 @@
...
+The parameter ``deactivated`` can be used to include deactivated users.
```
Comment (richvdh) [inline docs/admin_api/user_admin_api.rst]: likewise, more documentation please

File: docs/admin_api/user_admin_api.rst [hunk 15]
```diff
@@ -1,3 +1,46 @@
...
+            }, {
+                "name": "<user_id2>",
+                "password_hash": "<password_hash2>",
+                "is_guest": 0,
+                "admin": 1,
+                "user_type": null,
+                "deactivated": 0
+            }
+        ],
+        "next_token": 100
```
[the remark this answers is not in the document]
Comment (awesome-manuel) [reply docs/admin_api/user_admin_api.rst]: @richvdh do we need to take care of SQL injection here or is it enough to use `parse_string(request, ...)`?

File: changelog.d/5925.removal (hunk 1 above)
Comment (awesome-manuel) [reply changelog.d/5925.removal]: There is `changelog.d/5925.feature`. Also the v1/users endpoint does not support pagination.

File: docs/admin_api/user_admin_api.rst [hunk 16]
```diff
@@ -1,3 +1,46 @@
...
+=============
+
+This API returns all local user accounts.
+
+The api is::
+
+    GET /_synapse/admin/v2/users?offset=0&limit=10&guests=false
+
+including an ``access_token`` of a server admin.
+The parameters ``offset`` and ``limit`` are required only for pagination.
```
[the remark this answers is not in the document]
Comment (awesome-manuel) [reply docs/admin_api/user_admin_api.rst]: Also `name` -> `user_id`?

File: docs/admin_api/user_admin_api.rst (hunk 12 above)
Comment (awesome-manuel) [reply docs/admin_api/user_admin_api.rst]: @richvdh It is used in a `WHERE ? LIKE ?` expression in the SQL query. So `%` and `_` are valid regexp patterns. Do we need to take care of SQL injection for the filters or is it enough to use parse_string(request, ...)?

File: docs/admin_api/user_admin_api.rst (hunk 13 above)
Comment (awesome-manuel) [reply docs/admin_api/user_admin_api.rst]: You can only exclude guests using `guests=false` or `guests=0`.

File: synapse/storage/data_stores/main/__init__.py [hunk 17]
```diff
@@ -483,36 +481,58 @@ def get_users(self):
         return self._simple_select_list(
             table="users",
             keyvalues={},
-            retcols=["name", "password_hash", "is_guest", "admin", "user_type"],
+            retcols=[
```
Comment (awesome-manuel) [inline synapse/storage/data_stores/main/__init__.py]: Do we need to use `user_id` instead of `name` here as well?

**Comment by richvdh** [comment]
> Agreed. Would you prefer a fixup commit upon the existing ones, or changing the relevant commits but without rebase?

I'd prefer additional fixup commits. I'll squash them all down when I merge.

File: docs/admin_api/user_admin_api.rst (hunk 12 above)
Comment (richvdh) [reply docs/admin_api/user_admin_api.rst]: > It is used in a WHERE ? LIKE ? expression in the SQL query. So % and _ are valid regexp patterns

yup, can you add this to the documentation please

> Do we need to take care of SQL injection for the filters or is it enough to use parse_string(request, ...)?

I believe that as long as you use `?` parameters for the sql rather than inlining the parameters into the sql, you are safe from SQL injection.

File: docs/admin_api/user_admin_api.rst (hunk 13 above)
Comment (richvdh) [reply docs/admin_api/user_admin_api.rst]: again, can you make this clear in the documentation please. (I'd suggest picking one (`false`?) and documenting that as the correct format, even if you end up supporting `0` too)

File: synapse/storage/data_stores/main/__init__.py (hunk 17 above)
Comment (richvdh) [reply synapse/storage/data_stores/main/__init__.py]: this refers to the name of the column in the table, so no, you'll have to stick with `name`.

I'd prefer it if the response from the API used `user_id` instead of `name`, but I'm not going to insist on that.

File: docs/admin_api/user_admin_api.rst [hunk 18]
```diff
@@ -1,3 +1,47 @@
...
+This API returns all local user accounts.
+
+The api is::
+
+    GET /_synapse/admin/v2/users?from=0&limit=10&guests=false
+
+including an ``access_token`` of a server admin.
+The parameters ``from`` and ``limit`` are required only for pagination.
+By default, a ``limit`` of 100 is used.
+The parameter ``user_id`` can be used to filter by user id using an SQL regexp.
```
Comment (richvdh) [inline docs/admin_api/user_admin_api.rst]: regular expressions use syntax like `.` to mean "any character" and `.*` to mean "any number of any character".

I think this is a "SQL wildcard":

```suggestion
The parameter ``user_id`` can be used to filter by user id using an SQL wildcard.
```

File: synapse/storage/_base.py [hunk 19]
```diff
@@ -1577,39 +1582,51 @@ def _simple_select_list_paginate_txn(
...
         retcols,
+        filters=None,
+        keyvalues=None,
         order_direction="ASC",
     ):
         """
         Executes a SELECT query on the named table with start and limit,
         of row numbers, which may return zero or number of rows from start to limit,
         returning the result as a list of dicts.
+        Use filters to search attributes using SQL regular expressions and/or keyvalues
```
Comment (richvdh) [inline synapse/storage/_base.py]: backticks help show that we are referring to a specific parameter, rather than "filters" as a general concept

```suggestion

        Use `filters` to search attributes using SQL wildcards and/or `keyvalues`
```

File: docs/admin_api/user_admin_api.rst [hunk 20]
```diff
@@ -1,3 +1,47 @@
...
+default is to exclude deactivated users.
+If the endpoint does not return a ``next_token`` then there are no more users left.
+It returns a JSON body like the following:
+
+.. code:: json
+
+    {
+        "users": [
+            {
+                "user_id": "<user_id1>",
```
Comment (richvdh) [inline docs/admin_api/user_admin_api.rst]: are you sure it returns `user_id` rather than `name` here? I can't see anywhere in the code that it would return user_id.

File: docs/admin_api/user_admin_api.rst (hunk 18 above)
Comment (awesome-manuel) [reply docs/admin_api/user_admin_api.rst]: The SQL wildcards are actually just `%` or `_`, but some implementations like MS also use `*` and `?`. Futhermore I don't want to suggest to use those placeholders, the minimum this attribute should support with any database is a search for the given token.

File: docs/admin_api/user_admin_api.rst [hunk 21]
```diff
@@ -10,7 +10,8 @@ The api is::
 including an ``access_token`` of a server admin.
 The parameters ``from`` and ``limit`` are required only for pagination.
 By default, a ``limit`` of 100 is used.
-The parameter ``user_id`` can be used to filter by user id using an SQL regexp.
+The parameter ``user_id`` can be used to select only users with user ids that
+contain this value.
```
Comment (richvdh) [inline docs/admin_api/user_admin_api.rst]: I don't think this is what the implementation does, though.

If I make a request with `user_id=man`, it won't return `manuel`, even though `manuel` contains `man`.

File: docs/admin_api/user_admin_api.rst (hunk 20 above)
Comment (awesome-manuel) [reply docs/admin_api/user_admin_api.rst]: Ah right, I didn't look into the implementation of `_simple_search_list_txn` in detail. So I'm basically missing the wrapping `%`s to get the desired result...

File: synapse/storage/data_stores/main/__init__.py [hunk 22]
```diff
@@ -509,7 +509,7 @@ def get_users_paginate(
         """
         name_filter = {}
         if name:
-            name_filter["name"] = name
+            name_filter["name"] = "%%" + name + "%%"
```
Comment (richvdh) [inline synapse/storage/data_stores/main/__init__.py]: why `%%` rather than just `%` ?

Comment (awesome-manuel) [reply synapse/storage/data_stores/main/__init__.py]: copy & paste from `_simple_search_list_txn`

Comment (richvdh) [reply synapse/storage/data_stores/main/__init__.py]: like almost everything else in #1784, _simple_search_list_txn is doing it wrong.

a single % is fine here

**Comment by richvdh** [comment]
Thanks for your perseverence @awesome-manuel !

**Comment by awesome-manuel** [comment]
I definitely appreciate your input, @richvdh

**Comment by richvdh** [comment]
this seems to be broken on (at least) postgres: #6552. I guess we should have written some tests :/
